loopy/__init__.py

```diff
@@ -487,6 +488,68 @@ def make_copy_kernel(new_dim_tags, old_dim_tags=None):
 # }}}
 
 
+# {{{ einsum
+
+def make_einsum(spec, arg_names, **knl_creation_kwargs):
+    """Returns a :class:`LoopKernel` for evaluating array-based
+    operations using Einstein summation convention. Any
+    keyword arguments for kernel creation will be carried through
+    to :func:`make_kernel`.
+
+    :param spec: a string denoting the subscripts for
+        summation as a comma-separated list of subscript labels.
+        This follows the usual :func:`numpy.einsum` convention.
+        Note that the explicit indicator `->` for the precise output
+        form is required.
+    :param arg_names: an iterable of string types denoting
+        the names of the array operands.
+    """
+    arg_spec, out_spec = spec.split("->")
+    arg_specs = arg_spec.split(",")
+
+    if len(arg_names) != len(arg_specs):
+        raise ValueError(
+            f"Number of arg names ({arg_names}) should match the number "
+            f"of arg specs: {arg_specs}. Length of arg names is {len(arg_names)}; "
+            f"expecting {len(arg_specs)} arg names."
+        )
+
+    out_indices = set(out_spec)
+
+    all_indices = set(
+        idx
+        for argsp in arg_specs
+        for idx in argsp) | out_indices
+
+    sum_indices = all_indices - out_indices
+
+    from pymbolic import var
+    lhs = var("out")[tuple(var(i) for i in out_spec)]
+
+    rhs = 1
+    for arg_name, argsp in zip(arg_names, arg_specs):
+        rhs = rhs * var(arg_name)[tuple(var(i) for i in argsp)]
+
+    if sum_indices:
+        rhs = Reduction("sum", tuple(var(idx) for idx in sum_indices), rhs)
+
+    constraints = " and ".join(
+        "0 <= %s < N%s" % (idx, idx)
+        for idx in all_indices
+        )
+
+    if "name" not in knl_creation_kwargs:
+        knl_creation_kwargs["name"] = "einsum%dto%d_kernel" % (
+                len(all_indices), len(out_indices))
+
+    domain = "{[%s]: %s}" % (",".join(all_indices), constraints)
+    insns = [Assignment(lhs, rhs)]
+    return make_kernel(domain, insns,
+            **knl_creation_kwargs)
```

Comment:
```suggestion
    return make_kernel("{[%s]: %s}" % (",".join(all_indices), constraints), 
            [Assignment(lhs, rhs)], **knl_creation_kwargs)
```

Reply:
Done in https://github.com/inducer/loopy/pull/328/commits/b74a57c6722c0c3a68b60195aace857bfa098c86